compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -3248,7 +3248,20 @@ class TrackingTypeComparer(initctx: Context) extends TypeComparer(initctx) {
             MatchTypeTrace.noInstance(scrut, cas, fails)
             NoType
           case MatchResult.Reduced(tp) =>
-            tp.simplified
+            // A recursive match type will have the recursive call
+            // wrapped in a LazyRef.  For example in i18175, the recursive calls
+            // to IsPiped within the definition of IsPiped are all wrapped in LazyRefs.
+            // In addition to that, TypeMaps, such as the one that backs TypeOps.simplify,
+            // by default will rewrap a LazyRef when applying its function.
+            // The result of those two things means that given a big enough input
+            // that recursive enough times through one or multiple match types,
+            // reducing and simplifying the result of the case bodies,
+            // can end up with a large stack of directly-nested lazy refs.
+            // And if that nesting level breaches `Config.LogPendingSubTypesThreshold`,
+            // then TypeComparer will eventually start returning `false` for `isSubType`.
+            // Or, under -Yno-deep-subtypes, start throwing AssertionErrors.
+            // So, we eagerly strip that lazy ref here to avoid the stacking.
+            tp.stripLazyRef.simplified
```

Comment:
Have you tried calling stripLazyRef in simplified?

Reply:
Good idea.  It works.